cartographer/common/print_configuration_main.cc

```diff
@@ -0,0 +1,103 @@
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+
+DEFINE_string(configuration_directories, "",
+              "Comma separated list of directories in which configuration files"
+              " are searched, the last is always the Cartographer installation"
+              " to allow including files from there.");
+DEFINE_string(configuration_basename, "",
+              "Basename, i.e. not containing any directory prefix, of the "
+              "configuration file.");
+DEFINE_string(subdictionary, "",
+              "Only print a subdictionary referenced by its Lua ID, e.g.: "
+              "'--subdictionary trajectory_builder.trajectory_builder_3d'");
+
+namespace cartographer {
+namespace common {
+
+std::unique_ptr<LuaParameterDictionary> LoadLuaDictionary(
+    const std::vector<std::string>& configuration_directories,
+    const std::string& configuration_basename) {
+  auto file_resolver =
+      absl::make_unique<ConfigurationFileResolver>(configuration_directories);
+  const std::string code =
+      file_resolver->GetFileContentOrDie(configuration_basename);
+  // We use no reference count because we just want to print the configuration.
+  return LuaParameterDictionary::NonReferenceCounted(code,
+                                                     std::move(file_resolver));
+}
+
+void PrintSubdictionaryById(LuaParameterDictionary* lua_parameter_dictionary,
+                            const std::string& subdictionary_id) {
+  const std::vector<std::string> subdictionary_keys =
+      absl::StrSplit(subdictionary_id, '.', absl::SkipEmpty());
+  CHECK(!subdictionary_keys.empty()) << "Failed to parse 'subdictionary_id'.";
+  // Keep a stack to avoid memory glitches due to unique_ptr deletion.
+  std::vector<std::unique_ptr<LuaParameterDictionary>> stack;
+  for (const auto& key : subdictionary_keys) {
+    if (stack.empty()) {
+      stack.push_back(lua_parameter_dictionary->GetDictionary(key));
+      continue;
+    }
+    stack.push_back(stack.back()->GetDictionary(key));
+  }
+  CHECK(!stack.empty());
+  std::cout << subdictionary_id << " = " << stack.back()->ToString()
+            << std::endl;
+}
+
+}  // namespace common
+}  // namespace cartographer
+
+int main(int argc, char** argv) {
+  google::InitGoogleLogging(argv[0]);
+  google::SetUsageMessage(
+      "Resolves and compiles a Lua configuration and prints it to stdout.\n"
+      "The output can be restricted to a subdictionary using the optional "
+      "'--subdictionary' parameter, which can be given in Lua syntax.\n"
+      "The logs of the configuration file resolver are written to stderr if "
+      "'--logtostderr' is given.");
+  google::ParseCommandLineFlags(&argc, &argv, true);
+
+  if (FLAGS_configuration_directories.empty() ||
+      FLAGS_configuration_basename.empty()) {
+    google::ShowUsageWithFlagsRestrict(argv[0], "print_configuration_main");
+    return EXIT_FAILURE;
+  }
+
+  const std::vector<std::string> configuration_directories =
```

Comment:
nit: you can use here `std::vector<absl::string_view>` because of the `FLAGS_configuration_directories` lifetime.

Reply:
at line 55 as well